File: middlewares/dsh/core/protocol/host-bridge/src/codec/digest.ts
import { sha256 } from "@noble/hashes/sha2.js";
import { bytesToHex, concatBytes, utf8ToBytes } from "@noble/hashes/utils.js";
import type { GrantDigest, InputDigest, SchemaDigest } from "../contract/brands.js";
import { canonicalizeJson } from "./canonical.js";

const digest = <T extends string>(prefix: string, value: unknown): T => {
  const bytes = concatBytes(utf8ToBytes(prefix), canonicalizeJson(value));
  return `sha256:${bytesToHex(sha256(bytes))}` as T;
};

export const digestSchema = (value: unknown): SchemaDigest => digest<SchemaDigest>("muse-schema-v1\0", value);
export const digestInput = (value: unknown): InputDigest => digest<InputDigest>("muse-input-v1\0", value);
export const digestGrant = (value: unknown): GrantDigest => digest<GrantDigest>("muse-grant-v1\0", value);

File: middlewares/dsh/core/protocol/host-bridge/rust/src/digest.rs
use crate::json::{JsonBoundaryError, JsonLimits, canonicalize};
use serde_json::Value;
use sha2::{Digest, Sha256};

fn digest(prefix: &[u8], value: &Value, limits: JsonLimits) -> Result<String, JsonBoundaryError> {
    let canonical = canonicalize(value, limits)?;
    let mut hasher = Sha256::new();
    hasher.update(prefix);
    hasher.update(canonical);
    Ok(format!("sha256:{}", hex::encode(hasher.finalize())))
}

pub fn digest_schema(value: &Value, limits: JsonLimits) -> Result<String, JsonBoundaryError> {
    digest(b"muse-schema-v1\0", value, limits)
}

pub fn digest_input(value: &Value, limits: JsonLimits) -> Result<String, JsonBoundaryError> {
    digest(b"muse-input-v1\0", value, limits)
}

pub fn digest_grant(value: &Value, limits: JsonLimits) -> Result<String, JsonBoundaryError> {
    digest(b"muse-grant-v1\0", value, limits)
}
